### PR TITLE
Handle element library context

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
Addresses #678.

### CODE REVIEW

File: src/taipy/gui/gui.py [hunk 1]
```diff
@@ -1888,7 +1886,26 @@ def run(
...
+                if isinstance(lib, ElementLibrary):
+                    try:
+                        lib_context = lib.on_init(self)
+                        if isinstance(lib_context, tuple) and len(lib_context) > 1 and isinstance(lib_context[0], str) and lib_context[0].isidentifier():
+                            if lib_context[0] in glob_ctx:
+                                warnings.warn(f"Method {name}.on_init return a name already defined '{lib_context[0]}'.")
+                            else:
+                                glob_ctx[lib_context[0]] = lib_context[1]
+                        elif lib_context:
+                            warnings.warn(f"Method {name}.on_init should return a Tuple[str, Any] and the first element must be a valid Python identifier.")
```
Comment (FabienLelaquais) [inline src/taipy/gui/gui.py]: Missing () in method name reference.
"and the first element must be"  -> "where the first element must be"

File: src/taipy/gui/gui.py [hunk 2]
```diff
@@ -1888,7 +1886,26 @@ def run(
...
+
+        # call on_init on each library
+        for name, libs in self.__extensions.items():
+            for lib in libs:
+                if isinstance(lib, ElementLibrary):
+                    try:
+                        lib_context = lib.on_init(self)
+                        if isinstance(lib_context, tuple) and len(lib_context) > 1 and isinstance(lib_context[0], str) and lib_context[0].isidentifier():
+                            if lib_context[0] in glob_ctx:
+                                warnings.warn(f"Method {name}.on_init return a name already defined '{lib_context[0]}'.")
```
Comment (FabienLelaquais) [inline src/taipy/gui/gui.py]: Missing () in method name reference.
"return a name"  -> "returned a name"

File: src/taipy/gui/extension/library.py [hunk 3]
```diff
@@ -332,3 +332,16 @@ def get_data(self, library_name: str, payload: t.Dict, var_name: str, value: t.A
...
+    def on_init(self, gui: "Gui") -> t.Optional[t.Tuple[str, t.Any]]:
+        """
+        TODO
+        Called by `Gui.run()^`.
+
+        Arguments
+
+            gui: The `Gui^` instance.
+
+        Returns an optional Tuple composed of a context name (a valid Python identifier) and it's value
```
Comment (FabienLelaquais) [inline src/taipy/gui/extension/library.py]: ->
```
Returns:
    Returns an optional Tuple composed of a unique name for the library (that *must* be a Python identifier)  and its value.
```

File: src/taipy/gui/extension/library.py [hunk 4]
```diff
@@ -332,3 +332,16 @@ def get_data(self, library_name: str, payload: t.Dict, var_name: str, value: t.A
             value (any): The current value of the variable identified by *var_name*.
         """
         return None
+
+    def on_init(self, gui: "Gui") -> t.Optional[t.Tuple[str, t.Any]]:
+        """
+        TODO
+        Called by `Gui.run()^`.
+
+        Arguments
```
Comment (FabienLelaquais) [inline src/taipy/gui/extension/library.py:339]: =>
```
Arguments:
    gui: The `Gui^` instance.
```

File: src/taipy/gui/gui.py [hunk 5]
```diff
@@ -1888,7 +1886,26 @@ def run(
...
+                            else:
+                                glob_ctx[lib_context[0]] = lib_context[1]
+                        elif lib_context:
+                            warnings.warn(f"Method {name}.on_init should return a Tuple[str, Any] and the first element must be a valid Python identifier.")
+                    except Exception as e:  # pragma: no cover
+                        if not self._call_on_exception(f"{name}.on_init", e):
+                            warnings.warn(f"Method {name}.on_init raised an exception:\n{e}.")
+
+        # Initiate the Evalualtor with the right context
```
Comment (FabienLelaquais) [inline src/taipy/gui/gui.py]: Evalualtor -> Evaluator

File: src/taipy/gui/gui.py [hunk 6]
```diff
@@ -1888,7 +1886,26 @@ def run(
...
+                        lib_context = lib.on_init(self)
+                        if isinstance(lib_context, tuple) and len(lib_context) > 1 and isinstance(lib_context[0], str) and lib_context[0].isidentifier():
+                            if lib_context[0] in glob_ctx:
+                                warnings.warn(f"Method {name}.on_init return a name already defined '{lib_context[0]}'.")
+                            else:
+                                glob_ctx[lib_context[0]] = lib_context[1]
+                        elif lib_context:
+                            warnings.warn(f"Method {name}.on_init should return a Tuple[str, Any] and the first element must be a valid Python identifier.")
+                    except Exception as e:  # pragma: no cover
+                        if not self._call_on_exception(f"{name}.on_init", e):
```
Comment (FabienLelaquais) [inline src/taipy/gui/gui.py]: Missing () in method name reference. (twice?)

**Review by FabienLelaquais — APPROVED** [review]
Minor doc things.

File: src/taipy/gui/gui.py [hunk 7]
```diff
@@ -661,27 +661,25 @@ def __get_version(self) -> str:
...
-                                    if hasattr(elt, "_render_xhtml"):
-                                        elt_dict["render function"] = elt._render_xhtml.__code__.co_name
-                                    else:
-                                        elt_dict["react name"] = elt._get_js_name(element_name)
-                                    elts.append(elt_dict)
-            status.update({"libraries": libraries})
+        libraries: t.Dict[str, t.Any] = {}
+        for libs_list in self.__extensions.values():
+            for lib in libs_list:
+                if isinstance(lib, ElementLibrary):
```
Comment (dinhlongviolin1) [inline src/taipy/gui/gui.py]: Make it skip early so code is not nested.
If not instance -> continue

File: src/taipy/gui/gui.py [hunk 8]
```diff
@@ -661,27 +661,25 @@ def __get_version(self) -> str:
...
+                    libs = libraries.get(lib.get_name())
+                    if libs is None:
+                        libs = []
+                        libraries[lib.get_name()] = libs
+                    elts: t.List[t.Dict[str, str]] = []
+                    libs.append({"js module": lib.get_js_module_name(), "elements": elts})
+                    for element_name, elt in lib.get_elements().items():
+                        if isinstance(elt, Element):
```
Comment (dinhlongviolin1) [inline src/taipy/gui/gui.py]: Same thing

File: src/taipy/gui/gui.py [hunk 9]
```diff
@@ -1888,7 +1886,26 @@ def run(
         # base global ctx is TaipyHolder classes + script modules and callables
         glob_ctx: t.Dict[str, t.Any] = {t.__name__: t for t in _TaipyBase.__subclasses__()}
         glob_ctx.update({k: v for k, v in locals_bind.items() if inspect.ismodule(v) or callable(v)})
-        glob_ctx.update({Gui.__SELF_VAR: self})
+        glob_ctx[Gui.__SELF_VAR] = self
+
+        # call on_init on each library
+        for name, libs in self.__extensions.items():
+            for lib in libs:
+                if isinstance(lib, ElementLibrary):
```
Comment (dinhlongviolin1) [inline src/taipy/gui/gui.py]: Same thing

**Review by dinhlongviolin1 — APPROVED** [review]
Lgtm

**Pull request merged**